src/python/foglamp/data_purge/sqlalchemy_purge.py

```diff
@@ -0,0 +1,298 @@
+                    (datetime.datetime.strptime(age_timestamp, '%Y-%m-%d %H:%M:%S.%f') > last_connection):
+                # print('last_connection')
+                deleteQuery = sqlalchemy.delete(tableName).where(tableName.c.ts <= last_connection).where(
+                    tableName.c.id <= config['lastID'])
+                execute_delete(stmt=deleteQuery)
+
+            # All cases that do not fall under the IF statement uses age rather tha last_connection
+            else:
+                # print('age_timestamp')
+                # DELETE all rows that are older (or equal) to the age_timestamp.
+                #   DELETE FROM tableName WHERE ts <= age;
+                deleteQuery = sqlalchemy.delete(tableName).where(tableName.c.ts <= age_timestamp).where(
+                    tableName.c.id <= config['lastID'])
+                execute_delete(stmt=deleteQuery)
+
+        end = time.time()
+        end = datetime.datetime.fromtimestamp(end).strftime('%Y-%m-%d %H:%M:%S')
+
+        # Number of rows remain up to DELETE command
+        countAfter = get_count(countQuery)
+
+        # Add info to purgeStatus
+        purgeStatus['startTime'] = start
+        purgeStatus['complete'] = end
+
+        # rowsRemoved: How many rows were removed
+        purgeStatus['rowsRemoved'] = countBefore - countAfter
+
+        # unsentRowsRemoved: How many rows that were not sent to PI were removed
+        #   If config['retainUnsent'] is True,  then the value should be 0,  else it should be greater than 0
+        afterUnsent = get_count(unsentQuery)
+        purgeStatus['unsentRowsRemoved'] = beforeUnsent - afterUnsent
+
+        # failedRemovals: How many rows that were expected to get removed weren't removed
+        #   In cases where "config['retainUnsent']  is False and age_timestamp > last_connection" then failedRemovals
+        #   is the number of rows remaining that are <=  age_timestamp,  in all other cases,  the expected is 0.
+        if (config['retainUnsent'] is True) and \
+                (datetime.datetime.strptime(age_timestamp,  '%Y-%m-%d %H:%M:%S.%f') > last_connection):
+            failedToRemoveQuery = sqlalchemy.select([sqlalchemy.func.count()]).select_from(readings).where(
+                readings.c.ts <= last_connection)
+            purgeStatus['failedRemovals'] = get_count(failedToRemoveQuery)
+        else:
+            failedToRemoveQuery = sqlalchemy.select([sqlalchemy.func.count()]).select_from(readings).where(
+                readings.c.ts <= age_timestamp)
+            purgeStatus['failedRemovals'] = get_count(failedToRemoveQuery)
+
+        # rowsRemaining: How many rows remain remain up-to the delete
+        purgeStatus['rowsRemaining'] = countAfter
+
+        # The purgeStatus dictionary is stored into a dictionary containing all other purge runs,  where the key is
+        #   the `start` timestamp.
+        data[start] = purgeStatus
+
+        # Write to file
```

Comment:
This should all go in the Audit trail table as defined in the document https://docs.google.com/document/d/1GdMTerNq_-XQuAY0FJNQm09nbYQSjZudKJhY8x5Dq8c/edit#

Note the Audit trail table is called foglamp.log, it contains a JSONB column for this data.